File: lib/src/core/eval/detection_matcher.dart
import '../model/annotation.dart';
import '../model/coco_dataset.dart';
import '../model/detection_match.dart';
import '../model/eval_config.dart';
import '../model/model_run.dart';
import '../model/prediction.dart';
import 'iou.dart';

class DetectionMatcher {
  const DetectionMatcher();

  List<DetectionMatch> match({
    required CocoDataset dataset,
    required ModelRun modelRun,
    required EvalConfig config,
  }) {
    if (!config.classAwareMatching) {
      return _matchClassAgnostic(
        dataset: dataset,
        modelRun: modelRun,
        config: config,
      );
    }

    final List<DetectionMatch> matches = [];
    for (final int imageId in _allImageIds(dataset, modelRun)) {
      final List<GroundTruthAnnotation> groundTruths = _filteredGroundTruths(
        dataset.annotationsByImageId[imageId] ??
            const <GroundTruthAnnotation>[],
        config,
        matches,
      );
      final List<Prediction> predictions = _filteredPredictions(
        modelRun.predictionsByImageId[imageId] ?? const <Prediction>[],
        config,
      );

      final Set<int> categoryIds = {
        ...groundTruths.map((GroundTruthAnnotation gt) => gt.categoryId),
        ...predictions.map((Prediction pred) => pred.categoryId),
      };

      for (final int categoryId in categoryIds.toList()..sort()) {
        matches.addAll(
          _matchCategory(
            imageId: imageId,
            categoryId: categoryId,
            groundTruths: groundTruths
                .where(
                  (GroundTruthAnnotation gt) => gt.categoryId == categoryId,
                )
                .toList(),
            predictions: predictions
                .where((Prediction pred) => pred.categoryId == categoryId)
                .toList(),
            config: config,
          ),
        );
      }
    }
    return matches;
  }

  List<DetectionMatch> _matchCategory({
    required int imageId,
    required int categoryId,
    required List<GroundTruthAnnotation> groundTruths,
    required List<Prediction> predictions,
    required EvalConfig config,
  }) {
    final List<DetectionMatch> matches = [];
    final Set<int> matchedGtIndexes = {};
    final List<Prediction> sortedPredictions = [...predictions]
      ..sort(_comparePredictions);

    for (final Prediction prediction in sortedPredictions) {
      var bestUnmatchedIndex = -1;
      var bestUnmatchedIou = 0.0;
      var bestAnyMatched = false;
      var bestAnyIou = 0.0;

      for (var gtIndex = 0; gtIndex < groundTruths.length; gtIndex += 1) {
        final double iou =
            calculateIoU(prediction.bbox, groundTruths[gtIndex].bbox);
        if (iou > bestAnyIou) {
          bestAnyIou = iou;
          bestAnyMatched = matchedGtIndexes.contains(gtIndex);
        }
        if (!matchedGtIndexes.contains(gtIndex) && iou > bestUnmatchedIou) {
          bestUnmatchedIou = iou;
          bestUnmatchedIndex = gtIndex;
        }
      }

      if (bestUnmatchedIndex != -1 && bestUnmatchedIou >= config.iouThreshold) {
        matchedGtIndexes.add(bestUnmatchedIndex);
        matches.add(
          DetectionMatch(
            type: DetectionMatchType.truePositive,
            imageId: imageId,
            categoryId: categoryId,
            groundTruth: groundTruths[bestUnmatchedIndex],
            prediction: prediction,
            iou: bestUnmatchedIou,
            reason: DetectionMatchReason.matched,
          ),
        );
        continue;
      }

      final String reason = bestAnyMatched && bestAnyIou >= config.iouThreshold
          ? DetectionMatchReason.duplicatePrediction
          : (groundTruths.isEmpty
              ? DetectionMatchReason.noMatchingGroundTruth
              : DetectionMatchReason.lowIou);
      matches.add(
        DetectionMatch(
          type: DetectionMatchType.falsePositive,
          imageId: imageId,
          categoryId: categoryId,
          prediction: prediction,
          iou: bestAnyIou == 0 ? null : bestAnyIou,
          reason: reason,
        ),
      );
    }

    for (var gtIndex = 0; gtIndex < groundTruths.length; gtIndex += 1) {
      if (!matchedGtIndexes.contains(gtIndex)) {
        matches.add(
          DetectionMatch(
            type: DetectionMatchType.falseNegative,
            imageId: imageId,
            categoryId: categoryId,
            groundTruth: groundTruths[gtIndex],
            reason: DetectionMatchReason.missedGroundTruth,
          ),
        );
      }
    }
    return matches;
  }

  List<DetectionMatch> _matchClassAgnostic({
    required CocoDataset dataset,
    required ModelRun modelRun,
    required EvalConfig config,
  }) {
    final List<DetectionMatch> matches = [];
    for (final int imageId in _allImageIds(dataset, modelRun)) {
      final List<GroundTruthAnnotation> groundTruths = _filteredGroundTruths(
        dataset.annotationsByImageId[imageId] ??
            const <GroundTruthAnnotation>[],
        config,
        matches,
      );
      final List<Prediction> predictions = _filteredPredictions(
        modelRun.predictionsByImageId[imageId] ?? const <Prediction>[],
        config,
      )..sort(_comparePredictions);
      final Set<int> matchedGtIndexes = {};

      for (final Prediction prediction in predictions) {
        var bestIndex = -1;
        var bestIou = 0.0;
        for (var gtIndex = 0; gtIndex < groundTruths.length; gtIndex += 1) {
          if (matchedGtIndexes.contains(gtIndex)) {
            continue;
          }
          final double iou =
              calculateIoU(prediction.bbox, groundTruths[gtIndex].bbox);
          if (iou > bestIou) {
            bestIou = iou;
            bestIndex = gtIndex;
          }
        }
        if (bestIndex == -1 || bestIou < config.iouThreshold) {
          matches.add(
            DetectionMatch(
              type: DetectionMatchType.falsePositive,
              imageId: imageId,
              categoryId: prediction.categoryId,
              prediction: prediction,
              iou: bestIou == 0 ? null : bestIou,
              reason: DetectionMatchReason.noMatchingGroundTruth,
            ),
          );
          continue;
        }
        matchedGtIndexes.add(bestIndex);
        final GroundTruthAnnotation gt = groundTruths[bestIndex];
        matches.add(
          DetectionMatch(
            type: prediction.categoryId == gt.categoryId
                ? DetectionMatchType.truePositive
                : DetectionMatchType.falsePositive,
            imageId: imageId,
            categoryId: gt.categoryId,
            groundTruth: gt,
            prediction: prediction,
            iou: bestIou,
            reason: prediction.categoryId == gt.categoryId
                ? DetectionMatchReason.matched
                : DetectionMatchReason.wrongClass,
          ),
        );
      }

      for (var gtIndex = 0; gtIndex < groundTruths.length; gtIndex += 1) {
        if (!matchedGtIndexes.contains(gtIndex)) {
          matches.add(
            DetectionMatch(
              type: DetectionMatchType.falseNegative,
              imageId: imageId,
              categoryId: groundTruths[gtIndex].categoryId,
              groundTruth: groundTruths[gtIndex],
              reason: DetectionMatchReason.missedGroundTruth,
            ),
          );
        }
      }
    }
    return matches;
  }

  List<GroundTruthAnnotation> _filteredGroundTruths(
    List<GroundTruthAnnotation> groundTruths,
    EvalConfig config,
    List<DetectionMatch> matches,
  ) {
    if (!config.ignoreCrowd) {
      return [...groundTruths];
    }
    return groundTruths.where((GroundTruthAnnotation gt) {
      if (!gt.isCrowd) {
        return true;
      }
      matches.add(
        DetectionMatch(
          type: DetectionMatchType.ignored,
          imageId: gt.imageId,
          categoryId: gt.categoryId,
          groundTruth: gt,
          reason: DetectionMatchReason.ignoredCrowd,
        ),
      );
      return false;
    }).toList();
  }

  List<Prediction> _filteredPredictions(
    List<Prediction> predictions,
    EvalConfig config,
  ) {
    return predictions
        .where(
          (Prediction prediction) =>
              prediction.score >= config.confidenceThreshold,
        )
        .toList();
  }

  Set<int> _allImageIds(CocoDataset dataset, ModelRun modelRun) {
    return {
      ...dataset.imagesById.keys,
      ...modelRun.predictionsByImageId.keys,
    };
  }
}

int _comparePredictions(Prediction a, Prediction b) {
  final int scoreCompare = b.score.compareTo(a.score);
  if (scoreCompare != 0) {
    return scoreCompare;
  }
  final int categoryCompare = a.categoryId.compareTo(b.categoryId);
  if (categoryCompare != 0) {
    return categoryCompare;
  }
  final int xCompare = a.bbox.x.compareTo(b.bbox.x);
  if (xCompare != 0) {
    return xCompare;
  }
  final int yCompare = a.bbox.y.compareTo(b.bbox.y);
  if (yCompare != 0) {
    return yCompare;
  }
  final int widthCompare = a.bbox.width.compareTo(b.bbox.width);
  if (widthCompare != 0) {
    return widthCompare;
  }
  return a.bbox.height.compareTo(b.bbox.height);
}

File: server/cvmlab_server/core/detection_matcher.py
"""Detection matcher — port of lib/src/core/eval/detection_matcher.dart.

Reproduces the greedy, deterministic matching used by the Dart core so that
server-computed TP/FP/FN are identical to the local mode.
"""

from __future__ import annotations

from typing import List

from .iou import calculate_iou
from .models import (
    CocoDataset,
    DetectionMatch,
    EvalConfig,
    GroundTruthAnnotation,
    MatchReason,
    MatchType,
    ModelRun,
    Prediction,
)


def _prediction_sort_key(p: Prediction):
    # score desc, then category_id asc, then bbox x, y, width, height asc.
    return (-p.score, p.category_id, p.bbox.x, p.bbox.y, p.bbox.width, p.bbox.height)


def _all_image_ids(dataset: CocoDataset, model_run: ModelRun) -> List[int]:
    ids = set(dataset.images_by_id.keys()) | set(
        model_run.predictions_by_image_id.keys()
    )
    return list(ids)


def match(
    dataset: CocoDataset, model_run: ModelRun, config: EvalConfig
) -> List[DetectionMatch]:
    if not config.class_aware_matching:
        return _match_class_agnostic(dataset, model_run, config)

    matches: List[DetectionMatch] = []
    for image_id in _all_image_ids(dataset, model_run):
        ground_truths = _filtered_ground_truths(
            dataset.annotations_by_image_id.get(image_id, []), config, matches
        )
        predictions = _filtered_predictions(
            model_run.predictions_by_image_id.get(image_id, []), config
        )
        category_ids = sorted(
            {gt.category_id for gt in ground_truths}
            | {p.category_id for p in predictions}
        )
        for category_id in category_ids:
            matches.extend(
                _match_category(
                    image_id,
                    category_id,
                    [gt for gt in ground_truths if gt.category_id == category_id],
                    [p for p in predictions if p.category_id == category_id],
                    config,
                )
            )
    return matches


def _match_category(
    image_id: int,
    category_id: int,
    ground_truths: List[GroundTruthAnnotation],
    predictions: List[Prediction],
    config: EvalConfig,
) -> List[DetectionMatch]:
    matches: List[DetectionMatch] = []
    matched_gt: set = set()
    for prediction in sorted(predictions, key=_prediction_sort_key):
        best_unmatched_index = -1
        best_unmatched_iou = 0.0
        best_any_matched = False
        best_any_iou = 0.0
        for gt_index, gt in enumerate(ground_truths):
            iou = calculate_iou(prediction.bbox, gt.bbox)
            if iou > best_any_iou:
                best_any_iou = iou
                best_any_matched = gt_index in matched_gt
            if gt_index not in matched_gt and iou > best_unmatched_iou:
                best_unmatched_iou = iou
                best_unmatched_index = gt_index

        if best_unmatched_index != -1 and best_unmatched_iou >= config.iou_threshold:
            matched_gt.add(best_unmatched_index)
            matches.append(
                DetectionMatch(
                    type=MatchType.TRUE_POSITIVE,
                    image_id=image_id,
                    category_id=category_id,
                    ground_truth=ground_truths[best_unmatched_index],
                    prediction=prediction,
                    iou=best_unmatched_iou,
                    reason=MatchReason.MATCHED,
                )
            )
            continue

        if best_any_matched and best_any_iou >= config.iou_threshold:
            reason = MatchReason.DUPLICATE_PREDICTION
        elif not ground_truths:
            reason = MatchReason.NO_MATCHING_GROUND_TRUTH
        else:
            reason = MatchReason.LOW_IOU
        matches.append(
            DetectionMatch(
                type=MatchType.FALSE_POSITIVE,
                image_id=image_id,
                category_id=category_id,
                prediction=prediction,
                iou=None if best_any_iou == 0 else best_any_iou,
                reason=reason,
            )
        )

    for gt_index, gt in enumerate(ground_truths):
        if gt_index not in matched_gt:
            matches.append(
                DetectionMatch(
                    type=MatchType.FALSE_NEGATIVE,
                    image_id=image_id,
                    category_id=category_id,
                    ground_truth=gt,
                    reason=MatchReason.MISSED_GROUND_TRUTH,
                )
            )
    return matches


def _match_class_agnostic(
    dataset: CocoDataset, model_run: ModelRun, config: EvalConfig
) -> List[DetectionMatch]:
    matches: List[DetectionMatch] = []
    for image_id in _all_image_ids(dataset, model_run):
        ground_truths = _filtered_ground_truths(
            dataset.annotations_by_image_id.get(image_id, []), config, matches
        )
        predictions = sorted(
            _filtered_predictions(
                model_run.predictions_by_image_id.get(image_id, []), config
            ),
            key=_prediction_sort_key,
        )
        matched_gt: set = set()
        for prediction in predictions:
            best_index = -1
            best_iou = 0.0
            for gt_index, gt in enumerate(ground_truths):
                if gt_index in matched_gt:
                    continue
                iou = calculate_iou(prediction.bbox, gt.bbox)
                if iou > best_iou:
                    best_iou = iou
                    best_index = gt_index
            if best_index == -1 or best_iou < config.iou_threshold:
                matches.append(
                    DetectionMatch(
                        type=MatchType.FALSE_POSITIVE,
                        image_id=image_id,
                        category_id=prediction.category_id,
                        prediction=prediction,
                        iou=None if best_iou == 0 else best_iou,
                        reason=MatchReason.NO_MATCHING_GROUND_TRUTH,
                    )
                )
                continue
            matched_gt.add(best_index)
            gt = ground_truths[best_index]
            same = prediction.category_id == gt.category_id
            matches.append(
                DetectionMatch(
                    type=MatchType.TRUE_POSITIVE if same else MatchType.FALSE_POSITIVE,
                    image_id=image_id,
                    category_id=gt.category_id,
                    ground_truth=gt,
                    prediction=prediction,
                    iou=best_iou,
                    reason=MatchReason.MATCHED if same else MatchReason.WRONG_CLASS,
                )
            )

        for gt_index, gt in enumerate(ground_truths):
            if gt_index not in matched_gt:
                matches.append(
                    DetectionMatch(
                        type=MatchType.FALSE_NEGATIVE,
                        image_id=image_id,
                        category_id=gt.category_id,
                        ground_truth=gt,
                        reason=MatchReason.MISSED_GROUND_TRUTH,
                    )
                )
    return matches


def _filtered_ground_truths(
    ground_truths: List[GroundTruthAnnotation],
    config: EvalConfig,
    matches: List[DetectionMatch],
) -> List[GroundTruthAnnotation]:
    if not config.ignore_crowd:
        return list(ground_truths)
    result: List[GroundTruthAnnotation] = []
    for gt in ground_truths:
        if not gt.is_crowd:
            result.append(gt)
            continue
        matches.append(
            DetectionMatch(
                type=MatchType.IGNORED,
                image_id=gt.image_id,
                category_id=gt.category_id,
                ground_truth=gt,
                reason=MatchReason.IGNORED_CROWD,
            )
        )
    return result


def _filtered_predictions(
    predictions: List[Prediction], config: EvalConfig
) -> List[Prediction]:
    return [p for p in predictions if p.score >= config.confidence_threshold]
